dm/syncer/checkpoint.go

```diff
@@ -304,11 +351,64 @@ func NewRemoteCheckPoint(tctx *tcontext.Context, cfg *config.SubTaskConfig, id s
 		points:      make(map[string]map[string]*binlogPoint),
 		globalPoint: newBinlogPoint(binlog.NewLocation(cfg.Flavor), binlog.NewLocation(cfg.Flavor), nil, nil, cfg.EnableGTID),
 		logCtx:      tcontext.Background().WithLogger(tctx.L().WithFields(zap.String("component", "remote checkpoint"))),
+		snapshots:   make([]*remoteCheckpointSnapshot, 0),
+		snapshotSeq: 0,
 	}
 
 	return cp
 }
 
+// Snapshot make a snapshot of checkpoint and return the snapshot info.
+func (cp *RemoteCheckPoint) Snapshot(isSyncFlush bool) *SnapshotInfo {
+	cp.RLock()
+	defer cp.RUnlock()
+
+	// make snapshot is visit in single thread, so depend on rlock should be enough
+	cp.snapshotSeq++
+	id := cp.snapshotSeq
+
+	tableCheckPoints := make(map[string]map[string]tablePoint, len(cp.points))
+	for s, tableCps := range cp.points {
+		tableCpSnapshots := make(map[string]tablePoint)
+		for tbl, point := range tableCps {
+			if point.outOfDate() {
+				tableCpSnapshots[tbl] = point.savedPoint
+			}
+		}
+		if len(tableCpSnapshots) > 0 {
+			tableCheckPoints[s] = tableCpSnapshots
+		}
+	}
+
+	flushGlobalPoint := cp.globalPoint.outOfDate() || cp.globalPointSaveTime.IsZero() || (isSyncFlush && cp.needFlushSafeModeExitPoint)
```

Comment:
can you explain why we need `isSyncFlush && cp.needFlushSafeModeExitPoint` rather than `cp.needFlushSafeModeExitPoint` 😵‍💫

Reply:
yeah, as we discussed before, we only deal with SafeModeExitPoint for sync flush, while async flush only being used for routinely flush checkpoint. 

Quote from our prev discussion:
```
Plan to fix this issue, let me know if sounds good to you:

use cp.needFlushSafeModeExitPoint instead of snapshotCp.needFlushSafeModeExitPoint here
since flush safe mode exit point only happen for sync flush method, we only check if the current flush is sync flush
avoid to pass needFlushSafeModeExitPoint , instead we only add global point as part of snapshot when it pass check in Snapshot() creation method
Basically, these actions could simplify the mental model of flushing safe mode exit point here. safe mode exit point will only be flushed synchronously, and it only happen sequentially, so the worst case is we flush it twice(still doubt if that could ever happen), and we are also safe to flush it twice as it's done that incrementally.
```